crates/rust-analyzer/src/main_loop.rs

```diff
@@ -646,17 +644,6 @@ impl GlobalState {
                         this.loader.handle.invalidate(path.to_path_buf());
                     }
                 }
-
-                // Clear the diagnostics for the previously known version of the file.
-                // This prevents stale "cargo check" diagnostics if the file is
-                // closed, "cargo check" is run and then the file is reopened.
-                this.send_notification::<lsp_types::notification::PublishDiagnostics>(
```

Comment:
Yeah, I think we did this intentionally. Did the motivation that introduced this change still holds? Perhaps we need a config here?

Reply:
@kjeremy as an original author might know about motivation more and the config necessity, we'd better ask him.

I see zero reason to keep diagnostics cleared on close:
* next diagnostics publishing updates/erases old diagnostics from everywhere, including closed files
* it clearly breaks some workflows related to navigating from one error to another

Reply:
> next diagnostics publishing updates/erases old diagnostics from everywhere, including closed files, so we don't benefit anything clearing those diagnostics on a single file manually

If `cargo check` is still running when you open the file, you will see stale errors. This can be confusing.

Reply:
@kjeremy would you have a chance to take a look here? No worries if you can't!

Reply:
Are you thinking of https://github.com/rust-analyzer/rust-analyzer/issues/4266? That's still open.

Reply:
Ah, no, that's probably another race.